File: src/main/java/io/vproxy/vfx/ui/table/VTableView.java
package io.vproxy.vfx.ui.table;

import io.vproxy.vfx.control.scroll.NodeWithVScrollPane;
import io.vproxy.vfx.control.scroll.VScrollPane;
import io.vproxy.vfx.manager.font.FontManager;
import io.vproxy.vfx.manager.font.FontUsages;
import io.vproxy.vfx.manager.internal_i18n.InternalI18n;
import io.vproxy.vfx.theme.Theme;
import io.vproxy.vfx.util.FXUtils;
import javafx.beans.value.ChangeListener;
import javafx.collections.FXCollections;
import javafx.collections.ListChangeListener;
import javafx.collections.ObservableList;
import javafx.geometry.Pos;
import javafx.scene.control.Label;
import javafx.scene.layout.HBox;
import javafx.scene.layout.Pane;
import javafx.scene.layout.Region;
import javafx.scene.layout.VBox;
import javafx.scene.paint.Color;

import java.util.*;
import java.util.stream.Collectors;

public class VTableView<S> implements NodeWithVScrollPane {
    private final Pane root = new Pane();
    private final VBox rootVBox = new VBox();
    final HBox columnPane = new HBox();
    private final Pane fixColumnWidthColum = new Pane() {{
        setBackground(VTableColumn.BG);
        setPrefWidth(0);
    }};
    private final VScrollPane scrollPane = new VScrollPane();
    private final HBox dataPane = new HBox();
    private final Label emptyTableLabel = new Label(InternalI18n.get().emptyTableLabel()) {{
        FontManager.get().setFont(FontUsages.tableEmptyTableLabel, this);
        setTextFill(Theme.current().tableTextColor());
        setAlignment(Pos.CENTER);
        setTextFill(Color.GRAY);
    }};

    private final VTableSharedData<S> shared = new VTableSharedData<>(this);

    private final ObservableList<VTableColumn<S, ?>> columns = FXCollections.observableArrayList();
    private final ArrayList<VTableColumn<S, ?>> lastColumns = new ArrayList<>();
    final ObservableList<VTableRow<S>> items = FXCollections.observableArrayList();
    private final VTableRowListDelegate<S> itemsDelegate = new VTableRowListDelegate<>(items, shared);
    private VTableRow<S> selectedRow = null;

    public VTableView() {
        scrollPane.setContent(emptyTableLabel);
        columns.addListener(colsListener);
        items.addListener(itemsListener);

        root.getChildren().add(rootVBox);
        FXUtils.observeWidthHeightWithPreferred(root, rootVBox);

        var columnPane = new HBox();
        rootVBox.getChildren().addAll(columnPane, scrollPane.getNode());

        ChangeListener<? super Number> rootVBoxWidthChange = (ob, old, now) -> {
            if (now == null) return;
            scrollPane.getNode().setPrefWidth(now.doubleValue());
            if (rootVBox.getPrefWidth() != 0) {
                updateWidth(rootVBox.getPrefWidth());
            } else {
                updateWidth();
            }
            columnWidthFix();
        };
        rootVBox.widthProperty().addListener(rootVBoxWidthChange);
        rootVBox.prefWidthProperty().addListener(rootVBoxWidthChange);
        this.columnPane.widthProperty().addListener((ob, old, now) -> columnWidthFix());
        scrollPane.getNode().widthProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            emptyTableLabel.setPrefWidth(now.doubleValue());
            columnWidthFix();
        });
        scrollPane.getNode().heightProperty().addListener((ob, old, now) -> {
            if (now == null) return;
            emptyTableLabel.setPrefHeight(now.doubleValue() - 10);
        });
        rootVBox.heightProperty().addListener((ob, old, now) ->
            scrollPane.getNode().setPrefHeight(rootVBox.getHeight() - columnPane.getHeight()));
        columnPane.heightProperty().addListener((ob, old, now) ->
            scrollPane.getNode().setPrefHeight(rootVBox.getHeight() - columnPane.getHeight()));

        columnPane.getChildren().addAll(this.columnPane, fixColumnWidthColum);

        FXUtils.makeTopOnlyClipFor(columnPane, 4);
        FXUtils.makeBottomOnlyClipFor(dataPane, 4);
    }

    public Region getNode() {
        return root;
    }

    public ObservableList<VTableColumn<S, ?>> getColumns() {
        return columns;
    }

    public List<S> getItems() {
        return itemsDelegate;
    }

    public void setItems(List<S> items) {
        this.items.removeListener(itemsListener);
        this.items.clear();
        for (var item : items) {
            var row = new VTableRow<>(item, shared);
            row.setCols(columns);
            this.items.add(row);
        }
        this.items.addListener(itemsListener);
        sort();
        updateWidth();
        if (items.isEmpty()) {
            scrollPane.setContent(emptyTableLabel);
        } else {
            scrollPane.setContent(dataPane);
        }
    }

    @SuppressWarnings("FieldCanBeLocal")
    private final ListChangeListener<VTableColumn<S, ?>> colsListener = c -> {
        while (c.next()) {
            var added = c.getAddedSubList();
            var removed = c.getRemoved();

            for (int i = removed.size() - 1; i >= 0; --i) {
                var col = removed.get(i);
                int index = lastColumns.indexOf(col);
                lastColumns.remove(index);
                assert index != -1;
                for (var row : items) {
                    row.removeCol(index);
                }
                dataPane.getChildren().remove(index);
                columnPane.getChildren().remove(index);
                clearSort(col);
                col.shared = null;
            }
            for (var col : added) {
                var index = columns.indexOf(col);
                lastColumns.add(index, col);
                for (var row : items) {
                    row.addCol(index, col);
                }
                dataPane.getChildren().add(index, col.vbox);
                columnPane.getChildren().add(index, col.columnNode);
                col.shared = shared;
            }
        }
        updateWidth();
    };

    private final ListChangeListener<VTableRow<S>> itemsListener = c -> {
        while (c.next()) {
            var added = c.getAddedSubList();
            var removed = c.getRemoved();

            for (var row : removed) {
                row.remove();
            }
            for (var row : added) {
                row.setCols(columns);
                var index = items.indexOf(row);
                row.add(index);
            }
        }
        for (int i = 0; i < items.size(); i++) {
            var row = items.get(i);
            row.setBgColor(i);
        }
        if (items.isEmpty()) {
            scrollPane.setContent(emptyTableLabel);
        } else {
            scrollPane.setContent(dataPane);
            updateWidth();
        }
        var hasSort = false;
        for (var col : columns) {
            if (col.getSortPriority() > 0) {
                hasSort = true;
                break;
            }
        }
        if (hasSort) {
            sort();
        }
    };

    private void columnWidthFix() {
        double columnW = columnPane.getWidth();
        double scrollW = scrollPane.getNode().getWidth();
        if (scrollW > columnW) {
            fixColumnWidthColum.setPrefWidth(scrollW - columnW);
        } else {
            fixColumnWidthColum.setPrefWidth(0);
        }
    }

    void updateWidth() {
        updateWidth(scrollPane.getNode().getWidth());
    }

    private void updateWidth(double width) {
        if (width <= 0) return;
        if (columns.isEmpty()) return;
        var plan = buildUpdateWithPrefWidthPlan(width);
        if (plan != null) {
            updateWidth(plan);
            return;
        }
        plan = buildUpdateAvgConsiderMinMaxPlan(width);
        if (plan != null) {
            updateWidth(plan);
            return;
        }
        plan = buildAvgPlan(width);
        updateWidth(plan);
    }

    private Map<VTableColumn<S, ?>, Double> buildUpdateWithPrefWidthPlan(double width) {
        var ret = new HashMap<VTableColumn<S, ?>, Double>();

        var prefWCols = new ArrayList<VTableColumn<S, ?>>();

        double remain = width;
        int remainCnt = columns.size();
        for (var c : columns) {
            if (c.prefWidth > 0) {
                remain -= c.prefWidth;
                ret.put(c, c.prefWidth);
                prefWCols.add(c);
                --remainCnt;
            }
        }
        if (remain < 0) { // exceeds
            return null;
        }
        if (prefWCols.size() == columns.size()) {
            return ret;
        }

        return buildUpdateAvgConsiderMinMaxPlan(ret, prefWCols, remain, remainCnt);
    }

    private Map<VTableColumn<S, ?>, Double> buildUpdateAvgConsiderMinMaxPlan(double width) {
        var ret = new HashMap<VTableColumn<S, ?>, Double>();
        return buildUpdateAvgConsiderMinMaxPlan(ret, Collections.emptyList(), width, columns.size());
    }

    @SuppressWarnings("UnnecessaryContinue")
    private Map<VTableColumn<S, ?>, Double> buildUpdateAvgConsiderMinMaxPlan(
        HashMap<VTableColumn<S, ?>, Double> ret,
        List<VTableColumn<S, ?>> prefWCols,
        double remain, int remainCnt) {

        var exceedsMin = new ArrayList<VTableColumn<S, ?>>();
        var exceedsMax = new ArrayList<VTableColumn<S, ?>>();

        double avg = remain / remainCnt;
        while (true) {
            var exceedsMin0 = new ArrayList<VTableColumn<S, ?>>();
            var exceedsMax0 = new ArrayList<VTableColumn<S, ?>>();
            for (var c : columns) {
                if (prefWCols.contains(c)) continue;
                if (exceedsMin.contains(c)) continue;
                if (exceedsMax.contains(c)) continue;
                if (c.minWidth > 0) {
                    if (c.minWidth > avg) {
                        if (remain < c.minWidth) {
                            return null;
                        }
                        remain -= c.minWidth;
                        --remainCnt;
                        exceedsMin0.add(c);
                        ret.put(c, c.minWidth);
                        continue;
                    }
                }
                if (c.maxWidth > 0) {
                    if (c.maxWidth < avg) {
                        if (remain < c.maxWidth) {
                            return null;
                        }
                        remain -= c.maxWidth;
                        --remainCnt;
                        exceedsMax0.add(c);
                        ret.put(c, c.maxWidth);
                        continue;
                    }
                }
            }
            exceedsMin.addAll(exceedsMin0);
            exceedsMax.addAll(exceedsMax0);
            if (remainCnt == 0) {
                return ret;
            }
            avg = remain / remainCnt;
            if (exceedsMin0.size() == 0 && exceedsMax0.size() == 0) {
                for (var c : columns) {
                    if (prefWCols.contains(c)) continue;
                    if (exceedsMin.contains(c)) continue;
                    if (exceedsMax.contains(c)) continue;
                    ret.put(c, avg);
                }
                return ret;
            }
        }
    }

    private Map<VTableColumn<S, ?>, Double> buildAvgPlan(double width) {
        var avg = width / columns.size();
        var ret = new HashMap<VTableColumn<S, ?>, Double>();
        for (var c : columns) {
            ret.put(c, avg);
        }
        return ret;
    }

    private void updateWidth(Map<VTableColumn<S, ?>, Double> plan) {
        for (int i = 0; i < columns.size(); i++) {
            var c = columns.get(i);
            var w = plan.get(c);
            assert w != null;
            if (items.isEmpty()) {
                c.columnNode.setPrefWidth(w);
            }
            for (var row : items) {
                row.updateColWidth(i, w);
            }
        }
    }

    void refresh() {
        for (var c : columns) {
            c.vbox.getChildren().clear();
        }
        for (int i = 0; i < items.size(); i++) {
            var row = items.get(i);
            row.add();
            row.setBgColor(i);
        }
    }

    void updateRowNodeForColumn(VTableColumn<S, ?> col) {
        for (var row : items) {
            row.updateRowNodeForColumn(col);
        }
    }

    void selectRow(VTableRow<S> r) {
        if (selectedRow != null) {
            var index = items.indexOf(selectedRow);
            if (index != -1) {
                selectedRow.setSelected(false);
                selectedRow.setBgColor(index);
            }
        }
        r.setSelected(true);
        selectedRow = r;
    }

    public S getSelectedItem() {
        if (selectedRow == null) return null;
        if (!items.contains(selectedRow)) {
            return null;
        }
        return selectedRow.item;
    }

    public void sortBy(VTableColumn<S, ?> c, VTableSortOrder order) {
        if (c.comparator == null) {
            return;
        }
        if (c.getSortPriority() == 0) {
            int p = 0;
            for (var col : columns) {
                if (col.getSortPriority() > p) {
                    p = col.getSortPriority();
                }
            }
            ++p;
            c.setSort(p, order);
        } else {
            c.setSort(c.getSortPriority(), order);
        }
        sort();
    }

    public void clearSort(VTableColumn<S, ?> c) {
        c.resetSortOrder();
        if (c.getSortPriority() == 0) {
            return;
        }
        for (var col : columns) {
            if (col == c) continue;
            if (col.getSortPriority() < c.getSortPriority()) continue;
            col.decSortPriority();
        }
        c.resetSortPriority();
        sort();
    }

    void sort() {
        var cols = new ArrayList<>(columns).stream()
            .filter(c -> c.getSortPriority() > 0 && c.comparator != null)
            .sorted(Comparator.comparingInt(VTableColumn::getSortPriority))
            .collect(Collectors.toList());
        var tmp = new ArrayList<>(items);
        tmp.sort((a, b) -> {
            for (var c : cols) {
                var va = c.valueRetriever.apply(a.item);
                var vb = c.valueRetriever.apply(b.item);
                //noinspection unchecked,rawtypes
                var res = ((Comparator) c.comparator).compare(va, vb);
                if (res == 0) {
                    continue;
                }
                if (c.getSortOrder() == VTableSortOrder.ASC) {
                    return res;
                } else {
                    return -res;
                }
            }
            return Long.compare(a.rowId, b.rowId);
        });

        this.items.removeListener(itemsListener);
        this.items.clear();
        this.items.addAll(tmp);
        this.items.addListener(itemsListener);
        refresh();
    }

    @Override
    public VScrollPane getScrollPane() {
        return scrollPane;
    }

    @Override
    public Region getSelfNode() {
        return getNode();
    }
}

File: src-ts/src/ui/table/VTableView.ts
import { Node } from '../../javafx/Node.js';
import { Label } from '../../javafx/Label.js';
import { HBox } from '../../javafx/HBox.js';
import { VBox } from '../../javafx/VBox.js';
import { Pane } from '../../javafx/Pane.js';
import { ObservableList, ListChangeListener } from '../../javafx/Parent.js';
import { ChangeListener } from '../../javafx/Property.js';
import { Pos } from '../../javafx/layout.js';
import { Color } from '../../javafx/color.js';
import { VScrollPane } from '../../control/scroll/VScrollPane.js';
import type { NodeWithVScrollPane } from '../../control/scroll/NodeWithVScrollPane.js';
import { FontManager } from '../../manager/font/FontManager.js';
import { FontUsages } from '../../manager/font/FontUsages.js';
import { InternalI18n } from '../../manager/internal_i18n/InternalI18n.js';
import { Theme } from '../../theme/Theme.js';
import { FXUtils } from '../../util/FXUtils.js';
import { VTableColumn } from './VTableColumn.js';
import { VTableSortOrder } from './VTableSortOrder.js';
import { VTableSharedData } from './VTableSharedData.js';
import { VTableRow } from './VTableRow.js';
import { VTableRowListDelegate } from './VTableRowListDelegate.js';

export class VTableView<S> implements NodeWithVScrollPane {
  private readonly root: Pane = new Pane();
  private readonly rootVBox: VBox = new VBox();
  readonly columnPane: HBox = new HBox();
  private readonly fixColumnWidthColum: Pane = new Pane();
  private readonly scrollPane: VScrollPane = new VScrollPane();
  private readonly dataPane: HBox = new HBox();
  private readonly emptyTableLabel: Label = (() => {
    const label = new Label(InternalI18n.get().emptyTableLabel());
    FontManager.get().setFont(FontUsages.tableEmptyTableLabel, label);
    label.setTextFill(Theme.current().tableTextColor());
    label.setAlignment(Pos.CENTER);
    label.setTextFill(Color.GRAY);
    return label;
  })();

  private readonly shared: VTableSharedData<S> = new VTableSharedData<S>(this);

  private readonly columns: ObservableList<VTableColumn<S, unknown>> = new ObservableList<VTableColumn<S, unknown>>();
  private readonly lastColumns: VTableColumn<S, unknown>[] = [];
  readonly items: ObservableList<VTableRow<S>> = new ObservableList<VTableRow<S>>();
  private readonly itemsDelegate: VTableRowListDelegate<S> = new VTableRowListDelegate<S>(this.items, this.shared);
  private selectedRow: VTableRow<S> | null = null;

  constructor() {
    this.scrollPane.setContent(this.emptyTableLabel);
    this.columns.addListener(this.colsListener);
    this.items.addListener(this.itemsListener);

    this.root.getChildren().add(this.rootVBox);
    FXUtils.observeWidthHeightWithPreferred(this.root, this.rootVBox);

    const columnPane = new HBox();
    this.rootVBox.getChildren().addAll(columnPane, this.scrollPane.getNode());

    const rootVBoxWidthChange: ChangeListener<number> = (_old, now) => {
      if (now === null) return;
      this.scrollPane.getNode().setPrefWidth(now);
      if (this.rootVBox.getPrefWidth() !== 0) {
        this.updateWidthWith(this.rootVBox.getPrefWidth());
      } else {
        this.updateWidth();
      }
      this.columnWidthFix();
    };
    this.rootVBox.widthProperty.addListener(rootVBoxWidthChange);
    // Java registers the same listener on both widthProperty AND
    // prefWidthProperty — the latter fires synchronously when something
    // calls setPrefWidth, so column redistribution happens before the
    // actual rendered width propagates. Node now exposes prefWidthProperty
    // (kept in sync inside setPrefWidth), so we wire it up here for 1:1
    // parity with the Java original.
    this.rootVBox.prefWidthProperty.addListener(rootVBoxWidthChange);
    this.columnPane.widthProperty.addListener(() => this.columnWidthFix());
    this.scrollPane.getNode().widthProperty.addListener((_old, now) => {
      if (now === null) return;
      this.emptyTableLabel.setPrefWidth(now);
      this.columnWidthFix();
    });
    this.scrollPane.getNode().heightProperty.addListener((_old, now) => {
      if (now === null) return;
      this.emptyTableLabel.setPrefHeight(now - 10);
    });
    this.rootVBox.heightProperty.addListener(() =>
      this.scrollPane.getNode().setPrefHeight(this.rootVBox.getHeight() - this.columnPane.getHeight()),
    );
    this.columnPane.heightProperty.addListener(() =>
      this.scrollPane.getNode().setPrefHeight(this.rootVBox.getHeight() - this.columnPane.getHeight()),
    );

    columnPane.getChildren().addAll(this.columnPane, this.fixColumnWidthColum);

    this.fixColumnWidthColum.setBackground(VTableColumn.BG);
    this.fixColumnWidthColum.setPrefWidth(0);

    FXUtils.makeTopOnlyRoundedClipFor(columnPane, 4);
    FXUtils.makeBottomOnlyRoundedClipFor(this.dataPane, 4);
  }

  getNode(): Pane {
    return this.root;
  }

  getColumns(): ObservableList<VTableColumn<S, unknown>> {
    return this.columns;
  }

  getItems(): VTableRowListDelegate<S> {
    return this.itemsDelegate;
  }

  setItems(items: S[]): void {
    this.items.removeListener(this.itemsListener);
    this.items.clear();
    for (const item of items) {
      const row = new VTableRow<S>(item, this.shared);
      row.setCols(this.columns.toArray());
      this.items.add(row);
    }
    this.items.addListener(this.itemsListener);
    this.sort();
    this.updateWidth();
    if (items.length === 0) {
      this.scrollPane.setContent(this.emptyTableLabel);
    } else {
      this.scrollPane.setContent(this.dataPane);
    }
  }

  // Arrow-function fields so `this` is bound and the listener identity is
  // stable for add/removeListener (mirrors Java's final field lambdas).
  private readonly colsListener: ListChangeListener<VTableColumn<S, unknown>> = ({ added, removed }) => {
    // Process removals in reverse order so indices remain valid.
    for (let i = removed.length - 1; i >= 0; --i) {
      const col = removed[i]!;
      const index = this.lastColumns.indexOf(col);
      this.lastColumns.splice(index, 1);
      for (const row of this.items) {
        row.removeCol(index);
      }
      this.dataPane.getChildren().removeAt(index);
      this.columnPane.getChildren().removeAt(index);
      this.clearSort(col);
      col.shared = null;
    }
    for (const col of added) {
      const index = this.columns.indexOf(col);
      this.lastColumns.splice(index, 0, col);
      for (const row of this.items) {
        row.addCol(index, col);
      }
      this.dataPane.getChildren().insert(index, col.vbox);
      this.columnPane.getChildren().insert(index, col.columnNode);
      col.shared = this.shared;
    }
    this.updateWidth();
  };

  private readonly itemsListener: ListChangeListener<VTableRow<S>> = ({ added, removed }) => {
    for (const row of removed) {
      row.remove();
    }
    for (const row of added) {
      row.setCols(this.columns.toArray());
      const index = this.items.indexOf(row);
      row.addAt(index);
    }
    for (let i = 0; i < this.items.size(); i++) {
      const row = this.items.get(i);
      row.setBgColor(i);
    }
    if (this.items.isEmpty()) {
      this.scrollPane.setContent(this.emptyTableLabel);
    } else {
      this.scrollPane.setContent(this.dataPane);
      this.updateWidth();
    }
    let hasSort = false;
    for (const col of this.columns) {
      if (col.getSortPriority() > 0) {
        hasSort = true;
        break;
      }
    }
    if (hasSort) {
      this.sort();
    }
  };

  private columnWidthFix(): void {
    const columnW = this.columnPane.getWidth();
    const scrollW = this.scrollPane.getNode().getWidth();
    if (scrollW > columnW) {
      this.fixColumnWidthColum.setPrefWidth(scrollW - columnW);
    } else {
      this.fixColumnWidthColum.setPrefWidth(0);
    }
  }

  updateWidth(): void {
    this.updateWidthWith(this.scrollPane.getNode().getWidth());
  }

  private updateWidthWith(width: number): void {
    if (width <= 0) return;
    if (this.columns.isEmpty()) return;
    let plan = this.buildUpdateWithPrefWidthPlan(width);
    if (plan !== null) {
      this.applyWidthPlan(plan);
      return;
    }
    plan = this.buildUpdateAvgConsiderMinMaxPlan(width);
    if (plan !== null) {
      this.applyWidthPlan(plan);
      return;
    }
    plan = this.buildAvgPlan(width);
    this.applyWidthPlan(plan);
  }

  private buildUpdateWithPrefWidthPlan(width: number): Map<VTableColumn<S, unknown>, number> | null {
    const ret = new Map<VTableColumn<S, unknown>, number>();
    const prefWCols: VTableColumn<S, unknown>[] = [];

    let remain = width;
    let remainCnt = this.columns.size();
    for (const c of this.columns) {
      if (c.prefWidth > 0) {
        remain -= c.prefWidth;
        ret.set(c, c.prefWidth);
        prefWCols.push(c);
        --remainCnt;
      }
    }
    if (remain < 0) {
      return null;
    }
    if (prefWCols.length === this.columns.size()) {
      return ret;
    }

    return this.buildUpdateAvgConsiderMinMaxPlan2(ret, prefWCols, remain, remainCnt);
  }

  private buildUpdateAvgConsiderMinMaxPlan(width: number): Map<VTableColumn<S, unknown>, number> | null {
    const ret = new Map<VTableColumn<S, unknown>, number>();
    return this.buildUpdateAvgConsiderMinMaxPlan2(ret, [], width, this.columns.size());
  }

  private buildUpdateAvgConsiderMinMaxPlan2(
    ret: Map<VTableColumn<S, unknown>, number>,
    prefWCols: VTableColumn<S, unknown>[],
    remainIn: number,
    remainCntIn: number,
  ): Map<VTableColumn<S, unknown>, number> | null {
    const exceedsMin: VTableColumn<S, unknown>[] = [];
    const exceedsMax: VTableColumn<S, unknown>[] = [];

    let remain = remainIn;
    let remainCnt = remainCntIn;
    let avg = remain / remainCnt;
    while (true) {
      const exceedsMin0: VTableColumn<S, unknown>[] = [];
      const exceedsMax0: VTableColumn<S, unknown>[] = [];
      for (const c of this.columns) {
        if (prefWCols.includes(c)) continue;
        if (exceedsMin.includes(c)) continue;
        if (exceedsMax.includes(c)) continue;
        if (c.minWidth > 0) {
          if (c.minWidth > avg) {
            if (remain < c.minWidth) {
              return null;
            }
            remain -= c.minWidth;
            --remainCnt;
            exceedsMin0.push(c);
            ret.set(c, c.minWidth);
            continue;
          }
        }
        if (c.maxWidth > 0) {
          if (c.maxWidth < avg) {
            if (remain < c.maxWidth) {
              return null;
            }
            remain -= c.maxWidth;
            --remainCnt;
            exceedsMax0.push(c);
            ret.set(c, c.maxWidth);
            continue;
          }
        }
      }
      exceedsMin.push(...exceedsMin0);
      exceedsMax.push(...exceedsMax0);
      if (remainCnt === 0) {
        return ret;
      }
      avg = remain / remainCnt;
      if (exceedsMin0.length === 0 && exceedsMax0.length === 0) {
        for (const c of this.columns) {
          if (prefWCols.includes(c)) continue;
          if (exceedsMin.includes(c)) continue;
          if (exceedsMax.includes(c)) continue;
          ret.set(c, avg);
        }
        return ret;
      }
    }
  }

  private buildAvgPlan(width: number): Map<VTableColumn<S, unknown>, number> {
    const avg = width / this.columns.size();
    const ret = new Map<VTableColumn<S, unknown>, number>();
    for (const c of this.columns) {
      ret.set(c, avg);
    }
    return ret;
  }

  private applyWidthPlan(plan: Map<VTableColumn<S, unknown>, number>): void {
    for (let i = 0; i < this.columns.size(); i++) {
      const c = this.columns.get(i);
      const w = plan.get(c)!;
      if (this.items.isEmpty()) {
        c.columnNode.setPrefWidth(w);
      }
      for (const row of this.items) {
        row.updateColWidth(i, w);
      }
    }
  }

  refresh(): void {
    for (const c of this.columns) {
      c.vbox.getChildren().clear();
    }
    for (let i = 0; i < this.items.size(); i++) {
      const row = this.items.get(i);
      row.add();
      row.setBgColor(i);
    }
  }

  updateRowNodeForColumn(col: VTableColumn<S, unknown>): void {
    for (const row of this.items) {
      row.updateRowNodeForColumn(col);
    }
  }

  selectRow(r: VTableRow<S>): void {
    if (this.selectedRow !== null) {
      const index = this.items.indexOf(this.selectedRow);
      if (index !== -1) {
        this.selectedRow.setSelected(false);
        this.selectedRow.setBgColor(index);
      }
    }
    r.setSelected(true);
    this.selectedRow = r;
  }

  getSelectedItem(): S | null {
    if (this.selectedRow === null) return null;
    if (!this.items.contains(this.selectedRow)) {
      return null;
    }
    return this.selectedRow.item;
  }

  sortBy(c: VTableColumn<S, unknown>, order: VTableSortOrder): void {
    if (c.comparator === null) {
      return;
    }
    if (c.getSortPriority() === 0) {
      let p = 0;
      for (const col of this.columns) {
        if (col.getSortPriority() > p) {
          p = col.getSortPriority();
        }
      }
      ++p;
      c.setSort(p, order);
    } else {
      c.setSort(c.getSortPriority(), order);
    }
    this.sort();
  }

  clearSort(c: VTableColumn<S, unknown>): void {
    c.resetSortOrder();
    if (c.getSortPriority() === 0) {
      return;
    }
    for (const col of this.columns) {
      if (col === c) continue;
      if (col.getSortPriority() < c.getSortPriority()) continue;
      col.decSortPriority();
    }
    c.resetSortPriority();
    this.sort();
  }

  sort(): void {
    const cols = [...this.columns]
      .filter((c) => c.getSortPriority() > 0 && c.comparator !== null)
      .sort((a, b) => a.getSortPriority() - b.getSortPriority());
    const tmp = [...this.items];
    tmp.sort((a, b) => {
      for (const c of cols) {
        const va = c.valueRetriever(a.item);
        const vb = c.valueRetriever(b.item);
        const comparator = c.comparator!;
        const res = comparator(va, vb);
        if (res === 0) {
          continue;
        }
        if (c.getSortOrder() === VTableSortOrder.ASC) {
          return res;
        } else {
          return -res;
        }
      }
      // Java: Long.compare(a.rowId, b.rowId)
      if (a.rowId < b.rowId) return -1;
      if (a.rowId > b.rowId) return 1;
      return 0;
    });

    this.items.removeListener(this.itemsListener);
    this.items.clear();
    this.items.addAll(...tmp);
    this.items.addListener(this.itemsListener);
    this.refresh();
  }

  getScrollPane(): VScrollPane {
    return this.scrollPane;
  }

  getSelfNode(): Pane {
    return this.getNode();
  }
}
